### PR TITLE
Use SQLAlchemy idomatically and implement Response using specification.

### PR DESCRIPTION
**Pull request by andy-slac**
Use SQLAlchemy connection pool using MySQL config file.
Use SQLAlchemy text function for SQL.
Move engine to Flask app's config.

### CODE REVIEW

File: bin/metaServer.py [hunk 1]
```diff
@@ -31,12 +31,33 @@
...
+    db_config["username"] = db_config["user"]
+    del db_config["user"]
+    # SQLAlchemy part
+    url = URL("mysql",**db_config)
+    return sqlalchemy.create_engine(url)
+
+engine = initEngine()
+
+app.config["default_engine"] = engine
+
```
Comment (andy-slac) [inline bin/metaServer.py:60]: same comments as for dbserv/bin/dbServer.py

File: python/lsst/metaserv/metaREST_v0.py [hunk 2]
```diff
@@ -24,134 +24,134 @@
...
+        log.debug("Encountered an error processing request: '%s'" %(e.message))
+        response = _error(type(e).__name__, e.message)
+        status_code = INTERNAL_SERVER_ERROR
+    return _response(response, status_code)
+
+
+def _response(response, status_code):
+    fmt = request.accept_mimetypes.best_match(['application/json', 'text/html'])
+    if fmt == 'text/html':
+        response =  renderJsonResponse(response=response, status_code= status_code)
```
Comment (andy-slac) [inline python/lsst/metaserv/metaREST_v0.py]: extra space after = (x2)

File: python/lsst/metaserv/metaREST_v0.py [hunk 3]
```diff
@@ -24,134 +24,134 @@
...
+metaREST = Blueprint('metaREST', __name__, template_folder="templates")
 
 @metaREST.route('/', methods=['GET'])
 def getRoot():
     fmt = request.accept_mimetypes.best_match(['application/json', 'text/html'])
     if fmt == 'text/html':
-        return ("LSST Metadata Service v0 here. I currently support: "
-                "<a href='db'>/db</a> and <a href='image'>/image</a>.")
-    return "LSST Metadata Service v0 here. I currently support: /db and /image."
+        return ("LSST Metadata Service v0. See: <a href='db'>/db</a> and <a href='image'>/image</a>.")
```
Comment (andy-slac) [inline python/lsst/metaserv/metaREST_v0.py]: No parentheses needed here

File: python/lsst/metaserv/metaREST_v0.py [hunk 4]
```diff
@@ -24,134 +24,134 @@
...
-    return runDbQuery1(
-        "SHOW CREATE TABLE %s.%s" % (dbName, tableName),
-        None,
-        "Table '%s.%s'not found." % (dbName, tableName))
+    # Scalar
+    if SAFE_SCHEMA_PATTERN.match(dbName) and SAFE_TABLE_PATTERN.match(tableName):
+        query = "SHOW CREATE TABLE %s.%s" % (dbName, tableName)
+        message = "Table '%s.%s'not found." % (dbName, tableName)
+        return _resultOf(query, message=message)
+    return _response(_error("ValueError", "Database name or Table name is not safe"))
```
Comment (andy-slac) [inline python/lsst/metaserv/metaREST_v0.py]: `_response()` takes two arguments

File: python/lsst/metaserv/metaREST_v0.py [hunk 5]
```diff
@@ -24,134 +24,134 @@
...
+    try:
+        engine = current_app.config["default_engine"]
+        results = [[i for i in result] for result in engine.execute(query, **paramMap)]
+        if len(results):
+            response = _vector(results)
+        else:
+            response = _error(exception, message)
+            status_code = NOT_FOUND
+    except SQLAlchemyError as e:
+        log.debug("Encountered an error processing request: '%s'" %(e.message))
```
Comment (andy-slac) [inline python/lsst/metaserv/metaREST_v0.py]: logging-not-lazy

File: ups/metaserv.table [hunk 6]
```diff
@@ -1,4 +1,5 @@
 setupRequired(db)
```
Comment (andy-slac) [inline ups/metaserv.table:1]: +anaconda for sqlalchemy?

File: python/lsst/metaserv/metaREST_v0.py [hunk 7]
```diff
@@ -24,134 +24,134 @@
...
 
 @metaREST.route('/', methods=['GET'])
 def getRoot():
     fmt = request.accept_mimetypes.best_match(['application/json', 'text/html'])
     if fmt == 'text/html':
-        return ("LSST Metadata Service v0 here. I currently support: "
-                "<a href='db'>/db</a> and <a href='image'>/image</a>.")
-    return "LSST Metadata Service v0 here. I currently support: /db and /image."
+        return ("LSST Metadata Service v0. See: <a href='db'>/db</a> and <a href='image'>/image</a>.")
+    return "LSST Metadata Service v0. See: /db and /image."
```
Comment (andy-slac) [inline python/lsst/metaserv/metaREST_v0.py:52]: Is this string returned for application/json?

File: python/lsst/metaserv/metaREST_v0.py [hunk 8]
```diff
@@ -24,134 +24,134 @@
...
 def getImage():
     return ("meta/.../image not implemented. I am supposed to print list of "
             "supported image types here, something like: raw, template, coadd, "
             "jpeg, calexp, ... etc")
+
+_error = lambda exception, message: {"exception":exception, "message": message}
+_vector = lambda results: {"results": results}
+_scalar = lambda result: {"result": result}
+
+def _resultOf(query, paramMap=None, exception="NotFound",
```
Comment (andy-slac) [inline python/lsst/metaserv/metaREST_v0.py]: `_resultOf` and `_resultsOf` are almost identical, there is a lot of code duplication. Can you merge them into one method?

File: python/lsst/metaserv/metaREST_v0.py [hunk 9]
```diff
@@ -24,134 +24,134 @@
...
+              message="No Entity Found"):
+    status_code = OK
+    paramMap = paramMap if paramMap else {}
```
Comment (andy-slac) [inline python/lsst/metaserv/metaREST_v0.py]: Even shorter: `paramMap = paramMap or {}`

File: python/lsst/metaserv/metaREST_v0.py [hunk 10]
```diff
@@ -24,134 +24,134 @@
...
+    try:
+        engine = current_app.config["default_engine"]
+        result = engine.execute(query, **paramMap).first()
+        if result:
+            response = _scalar([i for i in result])
```
Comment (andy-slac) [inline python/lsst/metaserv/metaREST_v0.py]: `list(result)`?

File: python/lsst/metaserv/metaREST_v0.py [hunk 11]
```diff
@@ -24,134 +24,134 @@
...
 def getDbPerTypeDbNameTablesTableName(lsstLevel, dbName, tableName):
     '''Retrieves information about a table from a given database.'''
-    return runDbQuery1(
-        "SELECT DDT_Table.* FROM DDT_Table JOIN DbMeta USING (dbMetaId) "
-        "WHERE dbName=%s AND tableName=%s",
-        (dbName, tableName),
-        "Table '%s.%s'not found." % (dbName, tableName))
+    query = "SELECT DDT_Table.* FROM DDT_Table " \
+            + "JOIN DbMeta USING (dbMetaId) " \
+            + "WHERE dbName=:dbName AND tableName=:tableName"
```
Comment (andy-slac) [inline python/lsst/metaserv/metaREST_v0.py]: String literals can be concatenated without '+' (it's faster as it's done at compile time).

File: python/lsst/metaserv/metaREST_v0.py [hunk 12]
```diff
@@ -24,134 +24,134 @@
...
+    log.info(query)
+    paramMap = paramMap if paramMap else {}
+    try:
+        engine = current_app.config["default_engine"]
+        results = [[i for i in result] for result in engine.execute(query, **paramMap)]
+        if len(results):
+            response = _vector(results)
+        else:
+            response = _error(exception, message)
+            status_code = NOT_FOUND
```
Comment (andy-slac) [inline python/lsst/metaserv/metaREST_v0.py]: Hmm, for some queries empty result set could be normal. Do we need to distinguish between 'empty' and 'missing' resources?

Comment (brianv0) [inline python/lsst/metaserv/metaREST_v0.py]: In all cases in the above code, I communicate a message conveying that there are no objects found. In a few cases, those errors may be ambiguous ( as in getDbPerTypeDbNameTablesTableName), but we are aware of this and there's issue [DM-2537](https://jira.lsstcorp.org/browse/DM-2537) which specifically addresses this. 

I think DM-2537 should clear this up when I get to it.

**Pull request merged**